Fixes NPC Tucker for A Little Knowledge quest (Rolanberry to Vellum)

Aeltorian opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I understand I should leave resolving conversations to the LandSandBoat team so that reviewers won't miss what was said.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

This fixes the NPC Tucker to initiate a trade for Rolanberries to Vellum.

This PR fixes trading once again with NPC Tucker. 

 What was done; 

This is a fix for the quest A_Little_Knowledge which has an optional side trade quest with NPC Tucker.

trade code npc_utils was modified to where Tucker will only accept multiples of 12 up to 4 stacks as he states. Anymore and he'll only take the 4 stacks and return the rest. You still only have 3 attempts to trade him before your variable reaches the maximum and he'll be "out" of vellum. 

tucker's npc lua was also modified to add his 2 dialogs that he didn't have previously before. So if you talk to him outside of having the quest active, he'll talk about Rolanberries. If you try to trade him, he'll give a "what am I to do with these?" dialog. If quest is active and you trade him less than a stack, he will also give the "What am I to do with these?" dialog.

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

## Steps to test these changes

Trade NPC Tucker in Crawler's Nest (S) 4 stacks of Rolanberries, 3 times. You'll get a full stack of Vellum.

Aeltorian committed:
Fixes Rolanberry to Vellum Trade

This hotfixes the NPC Tucker to initiate a trade for Rolanberries to Vellum.

claywar commented on scripts/quests/crystalWar/A_Little_Knowledge.lua:
```diff
@@ -73,14 +73,11 @@ quest.sections =
                     -- In addition, need to determine if partial stacks are consumed and implement that if needed.
 
                     if
+                        npcUtil.tradeHas(trade, { { xi.items.ROLANBERRY, 12 } }) and
```
This is incorrect.  You can trade less and therefore get less.

Aeltorian replied:
> This is incorrect. You can trade less and therefore get less.

This directly conflicts with my personal testing as I could not trade less than 12, you should definitely get a "take your trading somewhere else" dialog. There's no reason to trade less than 12 anyways.

claywar replied:
Disregard, tradeHas, not tradeHasExactly

claywar replied:
Though, looking at this, it does check for >= 12, so you can't trade less.  Are you saying that you cannot trade anything that is not a full stack?

Aeltorian replied:
> Though, looking at this, it does check for >= 12, so you can't trade less. Are you saying that you cannot trade anything that is not a full stack?

You can trade 4 stacks and get 4 vellum. Trade 2 stacks, get 2 vellum. If it's not a full stack, he'll return those rolanberries to you. However, if you get a single piece of vellum during that trade round, you gain that 1 of 3 variables for a chance at the full vellum stack limit. You can't get 2.5 sheets of vellum. Whole numbers only.

claywar replied:
Ok, so it looks like the only thing that needs to change then is not eating the entire container, but instead confirming multiples of stacks

Aeltorian replied:
> Ok, so it looks like the only thing that needs to change then is not eating the entire container, but instead confirming multiples of stacks

I'm not sure what you mean.

claywar replied:
```lua
                    if
                        questOption >= 1 and
                        questOption <= 3 and
                        trade:hasItem(xi.items.ROLANBERRY) and
                        trade:getItemCount() >= 12 and
                        trade:hasItemQty(xi.items.ROLANBERRY, trade:getItemCount())
                    then
```
From above, you have to talk to Tucker first (option >= 1), you get 3 changes to trade (<= 3, and unavailable when the value is 4)

Trade can only have Rolanberries, and must have at least 12

```lua
local tuckerEventFinish = function(player, csid, option, npc)
    local numAwarded = quest:getLocalVar(player, 'numAwarded')

    if npcUtil.giveItem(player, { { xi.items.SHEET_OF_VELLUM, numAwarded } }) then
        player:confirmTrade()
        quest:incrementVar(player, 'Option', 1)
    end
end
```

Actually, we're doing this already.  We're only confirming `12 * numAwardedVellum`
```lua
                        local numAwardedVellum = math.min(4, math.floor(trade:getItemCount() / 12))

                        trade:confirmItem(xi.items.ROLANBERRY, numAwardedVellum * 12)
                        quest:setLocalVar(player, 'numAwarded', numAwardedVellum)
```

So the confirm only eats what we reward on.  What is the issue again?

claywar replied:
Mind you, I get the additional dialogue, and welcome it

Aeltorian replied:
trade:hasItem does not initiate on the server I play on. Which has the exact lua LSB has.

claywar replied:
Yep, that's the bug.  Needs a second parameter of 12, and then can remove the condition for `trade:getItemCount() >= 12 and`

claywar replied:
We don't want to use npcUtil with a static value of 12, because we're manually confirming later on in the code

Aeltorian replied:
> We don't want to use npcUtil with a static value of 12, because we're manually confirming later on in the code

So what should we do here? If I don't confirm the 12 here, then a Option variable would be counted if less than 12 was traded, I would think.

claywar replied:
I've already set confirming manually instead of relying on the npcUtil function
```lua
                        local numAwardedVellum = math.min(4, math.floor(trade:getItemCount() / 12))

                        trade:confirmItem(xi.items.ROLANBERRY, numAwardedVellum * 12)
                        quest:setLocalVar(player, 'numAwarded', numAwardedVellum)
```

claywar replied:
You can't enter that condition if less than 12 are traded (at least once the missing parameter is fixed)

Aeltorian replied:
> I've already set confirming manually instead of relying on the npcUtil function
> 
> ```lua
>                         local numAwardedVellum = math.min(4, math.floor(trade:getItemCount() / 12))
> 
>                         trade:confirmItem(xi.items.ROLANBERRY, numAwardedVellum * 12)
>                         quest:setLocalVar(player, 'numAwarded', numAwardedVellum)
> ```

So, remove the npc_utils line and keep the 

> if
>                         questOption >= 1 and
>                         questOption <= 3
> then

?

claywar replied:
```lua
                    if
                        questOption >= 1 and
                        questOption <= 3 and
                        trade:hasItem(xi.items.ROLANBERRY, 12) and
                        trade:hasItemQty(xi.items.ROLANBERRY, trade:getItemCount())
                    then
```
This verifies proper option, that there's at least 12 rolanberries, and the trade container only has rolanberries

Aeltorian replied:
Ok, let me test.

Aeltorian replied:
@claywar 

trade:hasItem refuses to work. Throwing a handling error with nil value.

claywar replied:
Sheesh, where was my head space that day (and today).  Change hasItem to hasItemQty

Aeltorian replied:
Yeah, this isn't working at all. NPC is not responding to the trade. Even removed the onTrade in npc lua too and still not responding...

claywar replied:
Give me a moment.  I'll check locally and get back.  You did talk to Tucker first, right? That's a requirement

Aeltorian replied:
> Give me a moment. I'll check locally and get back. You did talk to Tucker first, right? That's a requirement

yeah, It's not even giving me the correct dialogue CS neither. I've already resetted my variable. Still no change.

claywar replied:
```lua
                    if
                        questOption >= 1 and
                        questOption <= 3 and
                        trade:getItemQty(xi.items.ROLANBERRY) >= 12 and
                        trade:hasItemQty(xi.items.ROLANBERRY, trade:getItemCount())
                    then
 ```
Tested, and this performs as expected

Aeltorian replied:
Then I'm having an issue getting that variable back to 0.

Aeltorian replied:
recommiting. Works

Aeltorian committed:
Tucker Trade.getitemqty fixing

remodified the code to reflect the new changes to stray away from the npc_utils coding as suggested.

claywar commented on scripts/zones/Crawlers_Nest_[S]/npcs/Tucker.lua:
```diff
@@ -5,9 +5,11 @@
 local entity = {}
 
 entity.onTrade = function(player, npc, trade)
+    player:startEvent(14) -- doesn't want your trade dialog
```
Need some confirmation here.  If you're not on the quest, does this always happen?  If it doesn't, then it can move to an else block in the quest (I'll help to verify)

Aeltorian replied:
> Need some confirmation here. If you're not on the quest, does this always happen? If it doesn't, then it can move to an else block in the quest (I'll help to verify)

No, it does not. The Quest stuff overrides it.

claywar replied:
Right, but what happens if you trade to him when you're _not_ on the quest? This is the behavior I'm asking about here

claywar commented on scripts/zones/Crawlers_Nest_[S]/npcs/Tucker.lua:
```diff
@@ -5,9 +5,11 @@
 end
 
 entity.onTrigger = function(player, npc)
+    player:startEvent(15) -- rolanberry history dialog
```
This can probably move to default actions, but there's a run-up for other dialogues.  I want to say events 4 (one-time) and 5

Aeltorian replied:
> This can probably move to default actions, but there's a run-up for other dialogues. I want to say events 4 (one-time) and 5

I was JUST looking at that.

Aeltorian replied:
As we discussed that this would require a variable since one of these require a one time only event and we don't want to leave a forever variable in the system. So we'll move past this, I've also adjusted the event 14 in as a else catchall on trade during the quest as you verified on retail as it's a trade deny after the fact. Recommited. Lets get this working for those who need the vellum.

Aeltorian committed:
NPC Tucker dialogue adj to retail

readjusted Tuckers dialogue to mimic retail. 
Removed event 14 from tucker.lua and put it in quest line dialogue as it's where it belongs.

claywar commented on scripts/zones/Crawlers_Nest_[S]/npcs/Tucker.lua:
```diff
@@ -8,6 +8,7 @@ entity.onTrade = function(player, npc, trade)
 end
 
 entity.onTrigger = function(player, npc)
+    player:startEvent(15) -- rolanberry history dialog
```
I'm a bit wary putting this here. Is this his new action after completing the quest? If so, it would be safer to put this in a completed section of the quest and replaceDefault

Aeltorian replied:
Actually,  I agree with you. I'd like him to have some form of dialog when you talk to him before triggering this quest, else he'll say nothing when you talk to him. However. Player:startEvent may be the wrong dialog command to prompt it. I don't know all Lua just yet so the call to initiate dialog eludes me atm. Could you assist?

claywar replied:
I'm working on a PR right now so that we can handle the 4->5 transition.  4 will be the default action, 5 will replace after that has passed.  We're going to need to be careful with the priority tree for things, but it'll be covered soon.  For now, can we limit the scope of the PR to fixing the trade issue?

If 14/15 happen during the quest, I'm fine with keeping them there though, but not in the NPC script.

Aeltorian replied:
Alright, I'll remove stuffs from npc file and the quest lua should be all ready to go.

Aeltorian committed:
Tucker getting Quest to work

Deleting info the npc.lua

Aeltorian commented:
Alright, this latest commit. I can't verify if he continues to say the exact dialog;

"Fruits have an off-season, you know. What do you expect me to do with these? Eat 'em myself?"
"Go play at trading somewhere else."

After the tradings are complete, even when the quest is eventually completed, which ID 14 would have to go into the npc.lua as well.
